### PR TITLE
📎 Add more BorrowingMutex specification

### PR DESCRIPTION
**Pull request by MaciejNadolski98**

### CODE REVIEW

File: spec/truefi2/BorrowingMutex.spec [hunk 1]
```diff
@@ -1,11 +1,19 @@
 methods {
```
Comment (MaciejNadolski98) [inline spec/truefi2/BorrowingMutex.spec:1]: https://prover.certora.com/output/67403/ebde6252e5c64d2e3b74?anonymousKey=065a98a0d3a6587aba26572a0ae5b30dae6e898a

File: spec/truefi2/BorrowingMutex.spec [hunk 2]
```diff
@@ -23,13 +31,53 @@ rule functionDoesNotBanUnlockedBorrower() {
...
     calldataarg args;
     if (f.selector == lock(address, address).selector) {
-        address locker;
+        address locker = anyReasonableAddress();
         address lockee;
-        require isReasonableAddress(locker);
         lock(e, lockee, locker);
     } else {
-        sinvoke f(e, args);
+        f(e, args);
```
Comment (yuchenlintt) [inline spec/truefi2/BorrowingMutex.spec:38]: +1

For context: Shelly told us that `sinvoke` is default behavior and can be omitted unless we're doing fancy tricks.

File: spec/truefi2/BorrowingMutex.spec [hunk 3]
```diff
@@ -23,13 +31,53 @@ rule functionDoesNotBanUnlockedBorrower() {
...
+    require isBanned(borrower);
+
+    method f;
+    env e;
+    require isReasonableEnv(e);
+    calldataarg args;
+    f(e, args);
+
+    assert isBanned(borrower), "Borrower gets unbanned";
+}
```
Comment (yuchenlintt) [inline spec/truefi2/BorrowingMutex.spec:55]: What a great rule! If only all our rules could be this simple...

File: spec/truefi2/BorrowingMutex.spec [hunk 4]
```diff
@@ -1,11 +1,19 @@
...
     isBanned(address) returns bool envfree
+    locker(address) returns address envfree
 }
 
 definition isReasonableAddress(address alice) returns bool = alice != 0 && alice != 1;
 definition isReasonableEnv(env e) returns bool = isReasonableAddress(e.msg.sender);
 
+function anyReasonableAddress() returns address {
+    address alice;
+    havoc alice;
```
Comment (yuchenlintt) [inline spec/truefi2/BorrowingMutex.spec:12]: See discussion here: https://github.com/trusttoken/smart-contracts/pull/1108#discussion_r743292372.

I'm not sure whether the `havoc alice;` line is necessary.

**Review by yuchenlintt — APPROVED** [review]
LGTM!

These three new rules carry a lot of power :) You've just proven a big assumption (that our mutex implementation can prevent non-lockers from interfering in particular ways) that underlies how our FTL and LOC interact. Also defaults.

**Review by skalermo — APPROVED** [review]

**Pull request merged**